doc/tools/docgram/doc_grammar.ml

```diff
@@ -0,0 +1,1349 @@
+      edit_single_prod_r tl prods'' nt seen
+    | Sedit2 ("USE_NT", _) :: [] -> error "Bad position for USE_NT\n"; prods
+    | sym :: tl ->
+      edit_single_prod_r tl prods nt (sym :: seen)
+  in
+  edit_single_prod_r edit0 prods nt []
+
+let apply_edit_file g edits =
+  List.iter (fun b ->
+      let (nt, eprods) = b in
+      if not (edit_all_prods g nt eprods) then begin
+        let rec aux eprods prods add_nt =
+          match eprods with
+          | [] -> prods, add_nt
+          | (Snterm "DELETE" :: oprod) :: tl ->
+            aux tl (remove_prod oprod prods nt) add_nt
+          | (Snterm "DELETENT" :: _) :: tl ->  (* note this doesn't remove references *)
+            g.map <- NTMap.remove nt g.map;
+            g.order <- List.filter (fun nt1 -> nt1 <> nt) g.order;
+            aux tl prods false
+          | (Snterm "EDIT" :: oprod) :: tl ->
+            aux tl (edit_single_prod g oprod prods nt) add_nt
+          | (Snterm "REPLACE" :: oprod) :: (Snterm "WITH" :: rprod) :: tl ->
+            let posn = find_first oprod prods nt in
+            let prods' = match posn with
+            | None -> prods
+            | Some _ ->
+              let prods' = insert_after posn [rprod] prods in  (* insert new prod *)
+              remove_prod oprod prods' nt      (* remove orig prod *)
+            in
+            aux tl prods' add_nt
+          | (Snterm "REPLACE" :: _ as eprod) :: tl ->
+            error "Missing WITH after '%s' in '%s'\n" (prod_to_str eprod) nt;
+            aux tl prods add_nt
+          | prod :: tl ->
+            (* add a production *)
+            if has_match prod prods then
+              error "Duplicate production '%s' for %s\n" (prod_to_str prod) nt;
+            aux tl (prods @ [prod]) add_nt
+        in
+        let prods, add_nt =
+          aux eprods (try NTMap.find nt g.map with Not_found -> []) true in
+        if add_nt then begin
+          (* maybe add a new nt and its productions *)
+          if not (List.mem nt g.order) then
+            g.order <- g.order @ [nt];
+          g.map <- NTMap.add nt prods g.map
+        end
+      end)
+    edits
+
+
+(*** main routines ***)
+
+let nts_in_prod prod =
+  let rec traverse = function
+  | Sterm s -> []
+  | Snterm s -> [s]
+  | Slist1 sym
+  | Slist0 sym
+  | Sopt sym
+    -> traverse sym
+  | Slist1sep (sym, sep)
+  | Slist0sep (sym, sep)
+    -> traverse sym @ (traverse sep)
+  | Sparen sym_list -> List.concat (List.map traverse sym_list)
+  | Sprod sym_list_list -> List.concat (List.map (fun l -> List.concat (List.map traverse l)) sym_list_list)
+  | Sedit _
+  | Sedit2 _ -> []
+  in
+  List.rev (List.concat (List.map traverse prod)) (* preserve order, don't care about dups *)
+
+(* get the transitive closure of a non-terminal excluding "stops" symbols.
+   Preserve ordering to the extent possible *)
+let nt_closure g start stops =
+  let stop_set = StringSet.of_list stops in
+  let rec nt_closure_r res todo =
```

Comment:
The ordering matters for humans reading a grammar.  If I see the start symbol production:

```
start -> c b a
```

the printed grammar should show productions in the order `start c b a`.  IIRC StringSet orders its entries alphabetically, so it would show productions in the order `a b c start`, which is unfriendly.

The closure is useful when trying to break the grammar into sensible chunks (meaning you tweak the code locally to get what you want).  But a human-determined order will be better in the documentation, hence the use of `orderedGrammar`.